### PR TITLE
support multiple batch dimensions in Dense layer

### LINKED ISSUE #708: Plans for supporting higher dimensional data?
Currently most of Flux's models and functions are build for 2 dimensional data (the input dimension x the batch size) except for convolution. And as the slogan (Flux is the ML library that doesn't make you tensor) said, I guess it is intend to avoid higher dimensional data (the tensor). However, there are lot's of models that need (at least for the performance issue) those higher dimensional operations. So I just wondering about what is the Flux's plan for it, would we eventually go to somewhere like other DL frameworks do?

### PR DESCRIPTION
**Pull request by CarloLucibello**
Since most deep learning frameworks support it, we also should.

I can't find a corresponding issue. #282 is slightly related. 
After this, we should close #708 

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable
- [ ] Final review from `@dhairyagandhi96` (for API changes).

### CODE REVIEW

**Comment by DhairyaLGandhi** [comment]
Is this doing the correct thing here?

**Comment by CarloLucibello** [comment]
what do you mean? it does what is meant to do. Pytorch does the same thing. 
Notice that this is *not* #282 or #856

**Comment by mcabbott** [comment]
+1, this is the right behaviour.

I'd be tempted to make the existing method `x::AbstractVecOrMat` and then have `x::AbstractArray` which reshapes & calls that. But not a huge deal.

**Comment by CarloLucibello** [comment]
> I'd be tempted to make the existing method `x::AbstractVecOrMat` and then have `x::AbstractArray` which reshapes & calls that. But not a huge deal.

yeah, I didn't do it because the performance impact is negligible in any case, but we could. Anyway, as you said not a big deal

File: src/layers/basic.jl [hunk 1]
```diff
@@ -125,7 +130,11 @@ end
 
 function (a::Dense)(x::AbstractArray)
   W, b, σ = a.W, a.b, a.σ
-  σ.(W*x .+ b)
+  # reshape to handle dims > 1 has batch dimensions
```
Comment (CarloLucibello) [inline src/layers/basic.jl:133]:
```suggestion
  # reshape to handle dims > 1 as batch dimensions
```

File: src/layers/basic.jl [hunk 2]
```diff
@@ -85,14 +85,19 @@ extraChain(::Tuple{}, x) = ()
...
+Create a traditional `Dense` layer with in×out weight matrix `W` and 
+bias vector  `b` of length `out`. The forward pass is given by:
 
     y = σ.(W * x .+ b)
 
-The input `x` must be a vector of length `in`, or a batch of vectors represented
-as an `in × N` matrix. The out `y` will be a vector or batch of length `out`.
+The input `x` must be a vector of length `in`, a batch of vectors represented
+as an `in × N` matrix, or a higher order tensor where all dimensions
+after the first will be treated as batch dimensions.
```
Comment (CarloLucibello) [inline src/layers/basic.jl:95]:
```suggestion
after the first one will be treated as batch dimensions.
```

File: test/layers/basic.jl [hunk 3]
```diff
@@ -35,17 +35,23 @@ import Flux: activations
...
+      @test size(Dense(10, 5)(randn(10,2,3))) == (5,2,3)
+      @test size(Dense(10, 5)(randn(10,2,3,4))) == (5,2,3,4)
+    end
+    @testset "zeros" begin
+      @test Dense(10, 1, identity, initW=ones, initb=zeros)(ones(10,1)) == 10*ones(1, 1)
+      @test Dense(10, 1, identity, initW=ones, initb=zeros)(ones(10,2)) == 10*ones(1, 2)
+      @test Dense(10, 2, identity, initW=ones, initb=zeros)(ones(10,1)) == 10*ones(2, 1)
+      @test Dense(10, 2, identity, initW=ones, initb=zeros)([ones(10,1) 2*ones(10,1)]) == [10 20; 10 20]
+      @test Dense(10, 2, identity, initW=ones, bias=false)([ones(10,1) 2*ones(10,1)]) == [10 20; 10 20]
+    end
```
Comment (DhairyaLGandhi) [inline test/layers/basic.jl:54]:
```suggestion
    @testset "zeros" begin
      @test Dense(10, 1, identity, initW = ones, initb = zeros)(ones(10,1)) == 10*ones(1, 1)
      @test Dense(10, 1, identity, initW = ones, initb = zeros)(ones(10,2)) == 10*ones(1, 2)
      @test Dense(10, 2, identity, initW = ones, initb = zeros)(ones(10,1)) == 10*ones(2, 1)
      @test Dense(10, 2, identity, initW = ones, initb = zeros)([ones(10,1) 2*ones(10,1)]) == [10 20; 10 20]
      @test Dense(10, 2, identity, initW = ones, bias = false)([ones(10,1) 2*ones(10,1)]) == [10 20; 10 20]
    end
```

**Comment by CarloLucibello** [comment]
If we want to strictly adhere to ColPrac (and maybe we don't ), you should just approve Prs from collaborators with merge rights, then they merge by themselves:

## Reviewing, Approving and Merging PRs
- PRs must have 1 approval before they are merged.
- PR authors should not approve their own PRs.
- PRs should pass CI tests before being merged.
- PRs by people without merge rights must have approval from someone who has merge rights (who will usually then merge the PR).
- PRs by people with merge rights must have approval from someone else, who may or may not have merge rights (and then may merge their own PR).
- PRs by people with merge rights should not be merged by people other than the author (just approved).

**Comment by DhairyaLGandhi** [comment]
you should be able to use bors like usual

**Comment by CarloLucibello** [comment]
> you should be able to use bors like usual

sure, but I think that you need to hit Approve in Github's review (according to ColPrac)

**Review by DhairyaLGandhi — APPROVED** [review]

**Pull request merged**